src/cpu/cpu.cpp

```diff
@@ -131,6 +158,83 @@ void CPU_Core_Dynrec_Cache_Close();
 }
 #endif
 
+static void set_modern_cycles_config(const CpuMode mode)
+{
+	auto& conf = modern_cycles_config;
+
+	const auto new_cycles = [&]() -> std::optional<int> {
+		switch (mode) {
+		case CpuMode::Real: return conf.real_mode;
+		case CpuMode::Protected:
+			if (conf.protected_mode_auto) {
+				// 'cpu_cycles' controls both real and protected mode
+				return conf.real_mode;
+			} else {
+				return conf.protected_mode;
+			}
+		}
+		assert(false);
+		return {};
+	}();
+
+	const auto fixed_cycles = new_cycles.has_value();
+	if (fixed_cycles) {
+		if (conf.throttle) {
+			CPU_CycleAutoAdjust = true;
+			CPU_CyclePercUsed   = 100;
+			CPU_CycleLimit      = *new_cycles;
+		} else {
+			CPU_CycleAutoAdjust = false;
+			CPU_CycleMax        = *new_cycles;
+		}
+	} else {
+		// "max cycles" mode
+		CPU_CycleAutoAdjust = true;
+		CPU_CyclePercUsed   = 100;
+		CPU_CycleLimit      = -1;
+	}
+
+	CPU_Cycles    = 0;
+	CPU_CycleLeft = 0;
+
+	const auto real_mode_with_max_cycles = (mode == CpuMode::Real &&
+	                                        !fixed_cycles);
+	if (!real_mode_with_max_cycles) {
+		auto_determine_mode.auto_cycles = true;
+	}
+}
+
+void CPU_RestoreRealModeCyclesConfig()
+{
+	if (cpu.pmode || (!last_auto_determine_mode.auto_core &&
+	                  !last_auto_determine_mode.auto_cycles)) {
+		return;
+	}
+
+	auto_determine_mode      = last_auto_determine_mode;
+	last_auto_determine_mode = {};
+
+	if (auto_determine_mode.auto_cycles) {
+		if (legacy_cycles_mode) {
+			CPU_CycleAutoAdjust = false;
+			CPU_CycleLeft       = 0;
+			CPU_Cycles          = 0;
+			CPU_CycleMax        = old_cycle_max;
+		} else {
+			set_modern_cycles_config(CpuMode::Real);
+		}
+
+		GFX_NotifyCyclesChanged();
+	}
+#if (C_DYNAMIC_X86) || (C_DYNREC)
+	if (auto_determine_mode.auto_core) {
+		cpudecoder = &CPU_Core_Normal_Run;
```

Comment:
This always seemed odd to me (I know this is pre-existing behavior)  Why do we switch cores to normal in real mode?  Does dynrec have some issue with real mode specifically or is it more of "we don't need extra performance so let's use the simpler (possibly more accurate?) core"?  If the latter, would it not make more sense to switch to normal core depending on the number of cycles we are running?

Maybe this change is out of scope for this PR as it would probably require further testing but this behavior surprised me when I ran into it.  I was expecting auto to use the dynamic core all the time if it's available.

At the very least, it would be nice to document this.  Current help text on the config file is:

> #                 core: CPU core used in emulation ('auto' by default). 'auto' will switch to dynamic
> #                       if available and appropriate.

Maybe change that to "auto will switch to dynamic in protected mode if available" or something.

Reply:
> Does dynrec have some issue with real mode specifically or is it more of "we don't need extra performance so let's use the simpler (possibly more accurate?) core"? 

The dynarec core can sacrifice timing accuracy for speed. Self-modifying code can also fail on the dynamic core and run fine on the normal core.

> If the latter, would it not make more sense to switch to normal core depending on the number of cycles we are running?

Because of the above, I don't think we want to switch the core depending on the cycles value in use.

> At the very least, it would be nice to document this. Current help text on the config file is:

That's a great suggestion, and I started documenting the `cputype` and `core` settings, but that will be a separate follow-up PR. Let's not overload this one any further 😄